Typescript

morleyzhi opened the pull request:
Fixes #217.

Things to do before release:
- [x] Make sure prettier is running on src/
- [x] Upgrade to Webpack 4, since `ts-loader` won't run on Webpack 1
- [x] Make codebase pass tslint
- [x] Make sure tslint is running on src/ on commit
- [x] Update types, including #297 
- [x] Pass CI tests
- [x] Bump versions
- [x] Release the `typescript` branch as an alpha
- [x] Fix or document backwards-incompatible type definitions
- [x] Fix document generation
- [x] Upgrade in-development projects to typescript build
- [ ] Upgrade non-essential Stellar tools
- [x] Upgrade the Stellar Laboratory
- [ ] Once this has been up and used in production, merge into master

Akuukis commented:
Gave a try to upgrade stellar-sdk but not successful yet: see quick fixes at #329 , as well a start of type breaking changes I've found so far.

- `Server.*Record` -> `ServerApi.*Record`
- `Server.AccountResponse` -> `AccountResponse`
- `Server.*CallBuilder` -> `*CallBuilder`
- `HorizonResponseAccount` -> `Horizon.AccountResponse` (or `ServerApi.AccountRecord`)
- Still missing type for response of `fetch("https://horizon.stellar.org/")`

Akuukis commented:
@morleyzhi Commit d2fd175cd5772204c9b26e87a460aed98434ba04 introduced dublicate files `horizon_api.ts` and `horizon_api_types.ts`, which one do you want to keep?

Akuukis commented:
Next try to upgrade. Solved two more problem at #336
- `Omit` from lodash doesn't properly work expose extended interface props, so changed it to `Omit` from `utility-types`
- I've locally type-ized some error responses, so brought that into the library

### Currently outstanding errors

#### 1. dom specific types

Problem: `stellar-sdk` relies on dom-specific types, although it is not dom-specific itself.  
Workaround: every typescript consumer library will have to add "dom" to compilerOptions.  
Solution: We should refactor those away.  

```
../../node_modules/stellar-sdk/lib/call_builder.d.ts:7:36 - error TS2304: Cannot find name 'EventSource'.

7         EventSource: Constructable<EventSource>;
                                     ~~~~~~~~~~~

../../node_modules/stellar-sdk/lib/call_builder.d.ts:11:25 - error TS2304: Cannot find name 'MessageEvent'.

11     onmessage?: (event: MessageEvent) => void;
                           ~~~~~~~~~~~~

../../node_modules/stellar-sdk/lib/call_builder.d.ts:12:23 - error TS2304: Cannot find name 'MessageEvent'.

12     onerror?: (event: MessageEvent) => void;
                         ~~~~~~~~~~~~
```

#### 2. include of package.json

Problem: `stellar-sdk` imports `json` that requires special compilerOptions flag for all consumers.
Workaround: every typescript consumer library will have to add "dom" to compilerOptions.
Solution: don't use `import` (perhaps `require` or other way). Need to investigate how to resolve path correctly for both nodeJS imports and webpack.

```
../../node_modules/stellar-sdk/lib/index.d.ts:1:25 - error TS2732: Cannot find module '../package.json'. Consider using '--resolveJsonModule' to import module with '.json' extension

1 import { version } from "../package.json";
                          ~~~~~~~~~~~~~~~~~
```

morleyzhi commented:
@Akuukis I removed the extra file and just fixed the package.json imports.

Not sure how to address your point (1) about the DOM lib. Maybe it's reasonable to expect TS consumers to add that library for stellar-sdk. I'm happy to hear alternatives if you think of any.

Akuukis commented:
@morleyzhi I am one of the consumers that use stellar-sdk in nodejs. For me, keeping "dom" library out of Typescript helps to avoid accidentally using wrong types. 

I will check what it can be replaced with and then lets see if that's ok.

Akuukis commented:
@morleyzhi, please see PR #345 - `stellar-sdk` builds fine, and I'm keen to test it out from consumer project point of view.

morleyzhi commented:
@Akuukis Just published v2.0.0-beta.7.

Akuukis commented:
@morleyzhi great, I will continue tomorrow. Please note that a broken path within triple-slash reference is merged as part of #345 .

morleyzhi commented:
@Akuukis What's the remedy there? Your PR makes it seem that it has to be there, and it's purposefully broken, am I reading that incorrectly? Can we just remove that reference?

Akuukis commented:
@morleyzhi sorry for confusion, for that I put up `WIP:` prefix to title (but Github allowed you to merge anyways - did I need to write that in lowercase?). It was mostly for discussion whenever you are ok with this strategy. If we go this route, then we need the triple-slash to make consumers to discover and use the added `d.ts` file.

 Anyways, no, it is currently broken. I will look into it tomorrow.

Akuukis commented:
@morleyzhi I looked into it and AFAIK the best way is to fix dependencies instead of working around them. Here are PRs for that:
- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36259
- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36261

Once those are merged, I will adjust `stellar-sdk` appropriately.

Akuukis commented:
@morleyzhi I've finally successfully upgraded stellar-sdk to v2 for my codebase. Few little pieces along the way:
- https://github.com/stellar/js-stellar-sdk/pull/361
- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36614

There is still work to do with types, but as of now it LGTM to publish and improve them later.

Akuukis commented:
Correction: nodejs side works, but browser side requires another fix: https://github.com/stellar/js-stellar-sdk/pull/367
